File: src/main/java/de/espend/idea/php/annotation/doctrine/inspection/RepositoryClassInspection.java
package de.espend.idea.php.annotation.doctrine.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.*;
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocTag;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention;
import de.espend.idea.php.annotation.doctrine.util.DoctrineUtil;
import de.espend.idea.php.annotation.util.AnnotationUtil;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class RepositoryClassInspection extends LocalInspectionTool {
    public static final String MESSAGE = "[Annotations] Missing repository class";

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if(!DoctrineUtil.isDoctrineOrmInVendor(holder.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new MyAnnotationPropertyPsiElementVisitor("Doctrine\\ORM\\Mapping\\Entity") {
            @Override
            protected void visitAnnotationProperty(@NotNull PhpDocTag phpDocTag) {
                StringLiteralExpression repositoryClass = AnnotationUtil.getPropertyValueAsPsiElement(phpDocTag, "repositoryClass");
                if (repositoryClass == null) {
                    return;
                }

                if(!DoctrineUtil.repositoryClassExists(phpDocTag)) {
                    holder.registerProblem(
                        repositoryClass,
                        MESSAGE,
                        new DoctrineOrmRepositoryIntention()
                    );
                }
            }
        };
    }

    private static abstract class MyAnnotationPropertyPsiElementVisitor extends PsiElementVisitor {

        @NotNull
        private final String className;

        MyAnnotationPropertyPsiElementVisitor(@NotNull String className) {
            this.className = className;
        }

        @Override
        public void visitElement(@NotNull PsiElement psiElement) {
            if(!(psiElement instanceof PhpDocTag)) {
                super.visitElement(psiElement);
                return;
            }

            String name = ((PhpDocTag) psiElement).getName();
            if(AnnotationUtil.isBlockedAnnotationTag(name)) {
                super.visitElement(psiElement);
                return;
            }

            if(!AnnotationUtil.isAnnotationPhpDocTag((PhpDocTag) psiElement)) {
                super.visitElement(psiElement);
                return;
            }

            PhpClass phpClass = AnnotationUtil.getAnnotationReference(((PhpDocTag) psiElement));
            if(phpClass == null) {
                super.visitElement(psiElement);
                return;
            }

            if(!this.className.equals(phpClass.getPresentableFQN())) {
                super.visitElement(psiElement);
                return;
            }

            visitAnnotationProperty((PhpDocTag) psiElement);

            super.visitElement(psiElement);
        }

        abstract protected void visitAnnotationProperty(@NotNull PhpDocTag phpDocTag);
    }
}

File: src/main/kotlin/de/espend/idea/php/annotation/doctrine/inspection/RepositoryClassInspection.kt
package de.espend.idea.php.annotation.doctrine.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocTag
import de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention
import de.espend.idea.php.annotation.doctrine.util.DoctrineUtil
import de.espend.idea.php.annotation.util.AnnotationUtil

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class RepositoryClassInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!DoctrineUtil.isDoctrineOrmInVendor(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return object : MyAnnotationPropertyPsiElementVisitor("Doctrine\\ORM\\Mapping\\Entity") {
            override fun visitAnnotationProperty(phpDocTag: PhpDocTag) {
                val repositoryClass = AnnotationUtil.getPropertyValueAsPsiElement(
                    phpDocTag,
                    "repositoryClass",
                ) ?: return

                if (!DoctrineUtil.repositoryClassExists(phpDocTag)) {
                    holder.registerProblem(
                        repositoryClass,
                        MESSAGE,
                        DoctrineOrmRepositoryIntention(),
                    )
                }
            }
        }
    }

    private abstract class MyAnnotationPropertyPsiElementVisitor(
        private val className: String,
    ) : PsiElementVisitor() {
        override fun visitElement(psiElement: PsiElement) {
            if (psiElement !is PhpDocTag) {
                super.visitElement(psiElement)
                return
            }

            val name = psiElement.name
            if (AnnotationUtil.isBlockedAnnotationTag(name)) {
                super.visitElement(psiElement)
                return
            }

            if (!AnnotationUtil.isAnnotationPhpDocTag(psiElement)) {
                super.visitElement(psiElement)
                return
            }

            val phpClass = AnnotationUtil.getAnnotationReference(psiElement)
            if (phpClass == null) {
                super.visitElement(psiElement)
                return
            }

            if (className != phpClass.presentableFQN) {
                super.visitElement(psiElement)
                return
            }

            visitAnnotationProperty(psiElement)
            super.visitElement(psiElement)
        }

        protected abstract fun visitAnnotationProperty(phpDocTag: PhpDocTag)
    }

    companion object {
        const val MESSAGE = "[Annotations] Missing repository class"
    }
}
